app/api/src/docker/common/utils.js

```diff
@@ -35,7 +48,101 @@ async function runCommand(command, timeout = 5000) {
   })
 }
 
+function makeHash(script) {
+  return createHash('sha256').update(script).digest('hex')
+}
+
+async function checkIfAlreadyExists(params, s3) {
```

Comment:
You can create a signed url for a resource that doesn't exist (I guess aws assume you know if it does or not)
so we have to fetch the object's metadata/headObject to see if it exists, and we can grab the console message from metadata while we're at it.